### ISSUE TITLE
Preedit is not displayed when using IME on Neovide.

### ISSUE THREAD

**Issue by ZOI-dayo** [neovide/neovide#1931]
**Describe the bug**
Preedit is not displayed when using IME on Neovide.

https://github.com/neovide/neovide/assets/64856587/b812ff8c-ae13-446b-be8a-bad44219d259

**To Reproduce**
1. Open Neovide
2. Enter Japanese

**Expected behavior**
You should be able to see the string being converted even while you are typing using an IME.

↓ This is an example of expected behavior. (MacOS TextEdit)

https://github.com/neovide/neovide/assets/64856587/a5d2f097-6956-4a35-af32-da2dbb3177d5

**Screenshots**
There is a screen record in "Describe the bug" section.

**Desktop (please complete the following information):**
 - OS: macOS 13.3.1
 - Neovide Version : 372f29b79bebf2c2f40bf49bc2ad389ac9cf3369 (HEAD)
 - Neovim Version : NVIM v0.9.1

**Please run `neovide --log` and paste the contents of the `.log` file created in the current directory here:**
[neovide_rCURRENT.log](https://github.com/neovide/neovide/files/12014933/neovide_rCURRENT.log)

**Additional context**
I an using Google日本語入力(aka "mozc")
But the same bug occurs with the IME preinstalled in MacOS.

**Comment by fredizzimo** [comment]
I have an experimental branch of that here https://github.com/neovide/neovide/compare/main...fredizzimo:neovide:fsundvik/ime, but note that it's a bit behind neovide main right now, so for example the IME fixes for X11 are not included, so it does not work on X11. The other platforms should be fine though.

But it's not completely ready, selections does not work, transparency does not work, and it also does not deal with line wraps in any way.

I will finish it up, only after https://github.com/neovide/neovide/pull/1870, since the transparency fixes will be much easier with that.

**Comment by sevenc-nanashi** [comment]
Any updates on this issue? I'm facing to this problem too.

**Comment by fredizzimo** [comment]
I'm not sure if it's working, but I rebased https://github.com/neovide/neovide/compare/main...fredizzimo:neovide:fsundvik/ime

It should work on all platforms, but it's till missing proper transparency support, selections, and some kind of wrapping or scroll when the window edge is reached.

If someone wants to continue working on it feel free to notify me and take it over. I'm not sure when I will return to it yet.

**Comment by OliverChao** [comment]
https://github.com/neovide/neovide/compare/main...fredizzimo:neovide:fsundvik/ime doesn't work on MaxOS (M2)
I test it either as one single branch or as one merged branch.

https://github.com/neovide/neovide/pull/1959 works but have some problems: It covers the original text and lacks of word wrapping.

**Comment by OliverChao** [comment]
Hi @fredizzimo , I pushed one pr: https://github.com/neovide/neovide/pull/2198 to solve ime preedit missing problem.
But it needs some code review.
I will be grateful if you can check  my codes.

**Comment by micsama** [comment]
Hi,

Could you let me know if there are any updates or solutions for this issue now?

Thanks!

**Comment by fredizzimo** [comment]
I have been thinking about this, and I think most of the implementation should be implemented as a lua plugin on the Neovim side. A bit like standard code completion. That way all the word wrapping and scrolling when things goes out of the screen will work. It will also be possible to implement the correct undo behaviour.

Neovide could expose the Winit related [pre-edit events](https://docs.rs/winit/latest/winit/event/enum.Ime.html#:~:text=Describes%20input%20method%20events.%20This%20is%20also%20called%20a%20%E2%80%9Ccomposition)  to lua to make that possible.  

I have many other things to work, and I don't use IME myself, so I won't be doing it myself, but is there someone interested in working on this?

**Comment by kanium3** [comment]
Hi! I try to implement the feature showing preedit text on enabled IME. PR is #3110 .
The code to handle the IME was implemented as a Lua script that is loaded when neovim is first started. Therefore, since line breaks are handled by neovim, I don't think there will be any problems with rendering.
However, there is a separate issue to consider regarding possible cursor position.

**Comment by chyuaner** [comment]
I try to use macOS 15.4.1 Sequoia.
NeoVide installed with homebrew cask. 
NeoVide version is 0.15.0.

Have same issue.
Use macOS original IME Chinese is not showing.
Use McBopomofo IME is not showing too.

**Comment by fredizzimo** [comment]
* https://github.com/neovide/neovide/pull/3110 has now been merged.

Note that in itself it does not add pre-edit and has to be enabled, see https://neovide.dev/faq.html#how-to-enable-preedit-support-of-ime.

**Comment by fredizzimo** [comment]
I'm re-opening this since the PR had to be reverted https://github.com/neovide/neovide/pull/3218

It's being re-enabled here https://github.com/neovide/neovide/pull/3221, but it needs some support from Neovim, so no time estimates at the moment.

**Comment by Staten-Wang** [comment]
Does this mean that neovide can't handle Preedit displays at the moment, and is there a way to temporarily fix this?

**Comment by fredizzimo** [comment]
Yes, unfortunately it's not possible, it has been a known issue for a long time, with many different attempts to fix it. It's quite complex due to the many different modes, text scrolling and wrapping for example. But now I think we are on the right path, if we only can solve the Neovim issues.

For a workaround, you might be able to configure your IME to display the pre-edit in its own interface. For example, the Pinyn plugin for fcitx5 has a Predit mode setting. But I'm not sure if that works the same way as inline pre-edit would.

**Comment by Staten-Wang** [comment]
When is this problem expected to be resolved?

**Comment by LAST7** [comment]
> For a workaround, you might be able to configure your IME to display the pre-edit in its own interface.

A tip for anyone who wants to try this workaround, the config option for fcitx5 is at `~/.config/fctix5/config`:

```plaintext
[Behavior]
# Show preedit in application
PreeditEnabledByDefault=True
```

Also a hotkey for toggling it:

```plaintext
[Hotkey/TogglePreedit]
0=Control+Alt+P
```

P.S. Quite strange that `PreeditEnabledByDefault=True` results in disabling preedit. IDK why.

**Comment by falcucci** [comment]
the option to enable preedit text is available on nightly https://neovide.dev/faq.html#how-to-enable-preedit-support-of-ime
